drivers/iio/position/admt4000.c

```diff
@@ -0,0 +1,1572 @@
+}
+
+static int admt4000_get_hphase_config(struct admt4000_state *st, u8 hpha,
+			       u16 *pha)
+{
+	int ret;
+	int temp;
+
+	if (st->is_page_zero) {
+		ret = admt4000_set_page(st, false);
+		if (ret)
+			return ret;
+	}
+
+	switch (hpha) {
+	case 1:
+		ret = regmap_read(st->regmap, ADMT4000_02_REG_H1PH, &temp);
+		break;
+	case 2:
+		ret = regmap_read(st->regmap, ADMT4000_02_REG_H2PH, &temp);
+		break;
+	case 3:
+		ret = regmap_read(st->regmap, ADMT4000_02_REG_H3PH, &temp);
+		break;
+	case 8:
+		ret = regmap_read(st->regmap, ADMT4000_02_REG_H8PH, &temp);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	if (ret)
+		return ret;
+
+	*pha = FIELD_GET(ADMT4000_H_12BIT_PHA_MASK, temp);
+
+	return 0;
+}
+
+static int admt4000_get_temp(struct admt4000_state *st,
+		      bool is_primary)
+{
+	int ret;
+	int raw_temp;
+
+	if (!st->is_page_zero) {
+		ret = admt4000_set_page(st, true);
+		if (ret)
+			return ret;
+	}
+
+	if (is_primary)
+		ret = regmap_read(st->regmap, ADMT4000_00_REG_TMP0, &raw_temp);
+	else
+		ret = regmap_read(st->regmap, ADMT4000_00_REG_TMP1, &raw_temp);
+	if (ret)
+		return ret;
+
+	return FIELD_GET(ADMT4000_TEMP_MASK, raw_temp);
+}
+
+static int admt4000_get_raw_turns_and_angle(struct admt4000_state *st,
+				     int *turns, u16 *angle)
+{
+	int ret;
+	int temp, raw_turns;
+
+	ret = regmap_read(st->regmap, ADMT4000_AGP_REG_ABSANGLE, &temp);
+	if (ret)
+		return ret;
+	raw_turns = FIELD_GET(ADMT4000_TURN_CNT_MASK, temp); // 63
+	if (raw_turns > ADMT4000_TURN_CNT_THRES)
+		*turns = sign_extend64(raw_turns, 5);
+	else
+		*turns = raw_turns;
+	angle[0] = FIELD_GET(ADMT4000_ABS_ANGLE_ANGLE_MASK, temp);
+
+	ret = regmap_read(st->regmap, ADMT4000_AGP_REG_ANGLE, &temp);
+	angle[1] = FIELD_GET(ADMT4000_ANGLE_MASK, temp);
+
+	return 0;
+}
+
+static int admt4000_toggle_cnv(struct admt4000_state *st)
+{
+	int ret;
+	u32 temp;
+
+	if (st->cnv_gpio) {
+		gpiod_set_value_cansleep(st->cnv_gpio, 1);
+		udelay(1);
+		gpiod_set_value_cansleep(st->cnv_gpio, 0);
+	} else {
+		temp = FIELD_PREP(ADMT4000_CNV_EDGE_MASK, ADMT4000_RISING_EDGE);
+		ret = admt4000_reg_update(st, ADMT4000_AGP_REG_CNVPAGE,
+					  ADMT4000_CNV_EDGE_MASK, temp);
+
+		if (ret)
+			return ret;
+
+		temp = FIELD_PREP(ADMT4000_CNV_EDGE_MASK, ADMT4000_FALLING_EDGE);
+		ret =  admt4000_reg_update(st, ADMT4000_AGP_REG_CNVPAGE,
+					   ADMT4000_CNV_EDGE_MASK, temp);
+
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static irqreturn_t admt4000_interrupt(int irq, void *dev_id)
+{
+	struct iio_dev *indio_dev = dev_id;
+	struct admt4000_state *st = iio_priv(indio_dev);
+
+	if (iio_buffer_enabled(indio_dev))
+		iio_trigger_poll(st->trig);
+	else
+		complete(&st->completion);
+
+	return IRQ_HANDLED;
+};
+
+static irqreturn_t admt4000_trigger_handler(int irq, void *p)
+{
+	struct iio_poll_func *pf = p;
+	struct iio_dev *indio_dev = pf->indio_dev;
+	struct admt4000_state *st = iio_priv(indio_dev);
+	int ret;
+	int turns;
+	u16 angle[2];
+
+	mutex_lock(&st->lock);
```

Comment:
```suggestion
	guard(mutex)(&st->lock);
```
Then you don't need to call `mutex_unlock(&st->lock);` later.